src/main/java/org/takes/rq/RqLive.java

```diff
@@ -181,16 +174,20 @@ private static Integer legalCharacter(final Opt<Integer> data,
      * Obtains new byte if hasn't.
      * @param input Stream
      * @param data Empty or current data
+     * @param available Indicates whether or not it should check first if there
+     *  are available bytes
      * @return Next or current data
      * @throws IOException if input.read() fails
      */
     private static Opt<Integer> data(final InputStream input,
-        final Opt<Integer> data) throws IOException {
+        final Opt<Integer> data, final boolean available) throws IOException {
         final Opt<Integer> ret;
         if (data.has()) {
             ret = data;
+        } else if (available && input.available() <= 0) {
```

Comment:
@mkordas it is not easy to test it directly, but if you remove it you will currently get 6 failing tests so yes it is unit tested but indirectly. The other aspect to note is the fact that even if I refactored the code, I strictly respected the previous method flow because if you don't you will get failing tests.